[JSC] Support precise heap location analysis in CSE

hyjorc1 opened the pull request:
#### a7ad6d1e83ebee371e35dd4b62f74286558ef08e
<pre>
[JSC] Support precise heap location analysis in CSE
<a href="https://bugs.webkit.org/show_bug.cgi?id=285123">https://bugs.webkit.org/show_bug.cgi?id=285123</a>
<a href="https://rdar.apple.com/141970676">rdar://141970676</a>

Reviewed by NOBODY (OOPS!).

This patch enhances Common Subexpression Elimination (CSE) by
enabling precise heap tracking for non-typed JavaScript arrays
with constant index access. Previously, any write to an array
clobbered the entire abstract heap. Now, writes are tracked at
the per-cell level, improving CSE accuracy and reducing redundant
computations.

Benefits:
1. More precise invalidations for constant-indexed array writes.
2. Reduced recomputation, improving performance and resource efficiency.

Additionally, this patch optimizes trySetIndexQuickly by introducing
template parameter support for out-of-bound updates. Furthermore,
Copy-On-Write (COW) arrays are no longer set with the OutOfBounds
mode initially, enabling further CSE optimizations.

                            before                    after

loop-unrolling-4       31.4472+-0.1009           31.3783+-0.0931
loop-unrolling-5       35.5258+-0.1003     ^     32.7271+-0.0828        ^ definitely 1.0255x faste

* JSTests/microbenchmarks/loop-unrolling-4.js:
(test):
* JSTests/microbenchmarks/loop-unrolling-5.js: Added.
(assert):
(test):
* Source/JavaScriptCore/dfg/DFGCSEPhase.cpp:
* Source/JavaScriptCore/dfg/DFGClobberSet.cpp:
(JSC::DFG::addPreciseWrites):
* Source/JavaScriptCore/dfg/DFGClobberSet.h:
(JSC::DFG::PreciseClobberSet::overlaps):
(JSC::DFG::PreciseClobberSet::add):
(JSC::DFG::PreciseClobberSetAdd::PreciseClobberSetAdd):
(JSC::DFG::PreciseClobberSetAdd::operator() const):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
(JSC::DFG::preciseClobberize):
(JSC::DFG::PreciseWriteMethodClobberize::PreciseWriteMethodClobberize):
(JSC::DFG::PreciseWriteMethodClobberize::operator() const):
* Source/JavaScriptCore/dfg/DFGCommon.h:
* Source/JavaScriptCore/dfg/DFGHeapLocation.h:
(JSC::DFG::HeapLocation::hasConstantOffset const):
(JSC::DFG::HeapLocation::constantOffset const):
* Source/JavaScriptCore/dfg/DFGPreciseLocalClobberize.h:
(JSC::DFG::PreciseLocalClobberizeAdaptor::preciseWrite):
(JSC::DFG::preciseLocalClobberize):
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::putByVal):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::LLINT_SLOW_PATH_DECL):
* Source/JavaScriptCore/runtime/IndexingType.h:
* Source/JavaScriptCore/runtime/JSObject.h:
(JSC::JSObject::putByIndexInline):
(JSC::JSObject::trySetIndexQuickly): Deleted.
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::trySetIndexQuickly):
* Source/WTF/wtf/SetForScope.h:
(WTF::SetForScope::set):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a7ad6d1e83ebee371e35dd4b62f74286558ef08e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84669 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4394 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39057 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89808 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35720 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4483 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12368 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65918 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23737 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87714 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3406 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76951 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46191 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3285 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31163 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34795 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77632 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74153 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31965 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91183 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83711 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12008 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8912 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74386 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12235 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72755 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73511 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17874 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16315 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3456 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11960 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17400 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/106104 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11794 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25612 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15288 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13540 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on JSTests/microbenchmarks/loop-unrolling-4.js:
```diff
@@ -6,20 +6,29 @@ function assert(actual, expected) {
+    let a = [1, 2, 3, 4];
     let res = test(a);
     if (i == 0)
         expected = res;
```
How about adding yet another test file while keeping the original one as is?

Constellation commented on Source/JavaScriptCore/dfg/DFGCSEPhase.cpp:
```diff
@@ -140,6 +140,39 @@ class ImpureMap {
+    void clobber(const HeapLocation& location)
+    {
+        m_heapMap.removeIf([&](const std::unique_ptr<ImpureDataSlot>& slot) -> bool {
+#if !defined(NDEBUG)
```
#if ASSERT_ENABLED

Constellation commented on Source/JavaScriptCore/dfg/DFGClobberSet.h:
```diff
@@ -74,6 +74,30 @@ class ClobberSet {
+
+private:
+    HashSet<HeapLocation> m_clobberedLocations;
+    HashSet<Node*> m_clobberedBases;
```
This is not precisely trackable as it is possible that different Node* is pointing at the same base (aliasing).

hyjorc1 replied:
Thanks, updated. Current precise clobber is based on offset only.

Constellation reviewed (commented):
Why not just adding constant index as a supporting information, like what GetByOffset is doing? Keep in mind that we cannot rely on `base`'s value as it is possible that different base is pointing at the same object (aliasing).

hyjorc1 committed:
[JSC] Support precise heap location analysis in CSE
https://bugs.webkit.org/show_bug.cgi?id=285123
rdar://141970676

Reviewed by NOBODY (OOPS!).

This patch enhances Common Subexpression Elimination (CSE) by
enabling precise heap tracking for non-typed JavaScript arrays
with constant index access. Previously, any write to an array
clobbered the entire abstract heap. Now, writes are tracked at
the per-cell level, improving CSE accuracy and reducing redundant
computations.

Benefits:
1. More precise invalidations for constant-indexed array writes.
2. Reduced recomputation, improving performance and resource efficiency.

Additionally, this patch optimizes trySetIndexQuickly by introducing
template parameter support for out-of-bound updates. Furthermore,
Copy-On-Write (COW) arrays are no longer set with the OutOfBounds
mode initially, enabling further CSE optimizations.

                            before                    after

loop-unrolling-4       31.4472+-0.1009           31.3783+-0.0931
loop-unrolling-5       35.5258+-0.1003     ^     32.7271+-0.0828        ^ definitely 1.0255x faste

* JSTests/microbenchmarks/loop-unrolling-4.js:
(test):
* JSTests/microbenchmarks/loop-unrolling-5.js: Added.
(assert):
(test):
* Source/JavaScriptCore/dfg/DFGCSEPhase.cpp:
* Source/JavaScriptCore/dfg/DFGClobberSet.cpp:
(JSC::DFG::addPreciseWrites):
* Source/JavaScriptCore/dfg/DFGClobberSet.h:
(JSC::DFG::PreciseClobberSet::overlaps):
(JSC::DFG::PreciseClobberSet::add):
(JSC::DFG::PreciseClobberSetAdd::PreciseClobberSetAdd):
(JSC::DFG::PreciseClobberSetAdd::operator() const):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
(JSC::DFG::preciseClobberize):
(JSC::DFG::PreciseWriteMethodClobberize::PreciseWriteMethodClobberize):
(JSC::DFG::PreciseWriteMethodClobberize::operator() const):
* Source/JavaScriptCore/dfg/DFGCommon.h:
* Source/JavaScriptCore/dfg/DFGHeapLocation.h:
(JSC::DFG::HeapLocation::hasConstantOffset const):
(JSC::DFG::HeapLocation::constantOffset const):
* Source/JavaScriptCore/dfg/DFGPreciseLocalClobberize.h:
(JSC::DFG::PreciseLocalClobberizeAdaptor::preciseWrite):
(JSC::DFG::preciseLocalClobberize):
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::putByVal):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::LLINT_SLOW_PATH_DECL):
* Source/JavaScriptCore/runtime/IndexingType.h:
* Source/JavaScriptCore/runtime/JSObject.h:
(JSC::JSObject::putByIndexInline):
(JSC::JSObject::trySetIndexQuickly): Deleted.
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::trySetIndexQuickly):
* Source/WTF/wtf/SetForScope.h:
(WTF::SetForScope::set):
